src/renderer/components/+catalog/catalog-entity-details.tsx

```diff
@@ -75,7 +75,7 @@ export class CatalogEntityDetails<T extends CatalogEntity> extends Component<Pro
                     item.onRun(catalogEntityRunContext);
                   }
 
-                  if (typeof item.entity.onClickDetailIcon === "function" && Boolean(item.entity.metadata?.labels?.triggerOnRunAfterOnClickDetailIcon)) {
+                  if (typeof item.entity.onClickDetailIcon === "function" && item.entity.spec?.triggerOnRunAfterOnClickDetailIcon) {
                     item.onRun(catalogEntityRunContext);
```

Comment:
1. This should probably be in the above `if` so as to not check twice.
2. I don't think this does what you want it to do because `onClickDetailIcon` is very much an asyncronous function which will return immediately. I think it would be better to have a "beforeOnRun" hook that gets passed a type that can be `preventDefault`-ed.

Reply:
1. It's a bit hard to explain why order like this, maybe it's easier to look at the test case "prioritize onClickDetailIcon over onRun, but always trigger onRun if triggerOnRunAfterOnClickDetailIcon='true'" and "prioritize onClickDetailIcon over onRun"

"prioritize onClickDetailIcon over onRun" is a test case for
```tsx
if (typeof item.entity.onClickDetailIcon === "function") {
    item.onClickDetailIcon(catalogEntityRunContext);
  }	else {
    item.onRun(catalogEntityRunContext);
  }
```
and
"prioritize onClickDetailIcon over onRun, but always trigger onRun if triggerOnRunAfterOnClickDetailIcon='true'" is for this part
```tsx
if (typeof item.entity.onClickDetailIcon === "function" && item.entity.spec?.triggerOnRunAfterOnClickDetailIcon) {
    item.onRun(catalogEntityRunContext);
  }
```

2. `onClickDetailIcon` is a sync function (see typings) that doesn't return. What you describe: "beforeOnRun" + preventDefaulted equals to having `onClickDetailIcon` and set `triggerOnRunAfterOnClickDetailIcon=false`

Reply:
It is a sync function yes, but you are doing an "emit" which, while sync. Returns immediately and not after all the listeners return.

Reply:
Not sure I understand, can you be elaborate, like which functions return?

Reply:
Take https://github.com/lensapp/lens/pull/3911/files/aa70ca4154f945c0ee369167e4826248108856b7#diff-5099912721d6845c3d29e2acaa3fd4a50285eec049fd7aa76e8f025e363b4b2bR49-R53 for instance, any method that has been registered with the Weblinks category will be added to the JS event loop and then the `emit` call will return immediately but because of how the event loop works the return will happen and then continue executing before any of those listeners (or handlers) are executed.

Reply:
That's not true, at least for Node.js events and the "events" package for npm. (Not sure which eventemitter we're using here).

```
var EventEmitter = require('events')
 
var ee = new EventEmitter()
ee.on('message', function () {
  console.log("2");
})
console.log("1");
ee.emit('message', 'hello world')
console.log("3");
```

This prints 1, 2 and then 3.

Node.js docs: "When the EventEmitter object emits an event, all of the functions attached to that specific event are called synchronously. Any values returned by the called listeners are ignored and discarded."

Reply:
okay fair enough. However async function return immediately anyway so if any they aren't awaited.

Reply:
Not sure about that too, but I guess it's out of scope of this PR..